### PR TITLE
Ctor 665 hpe primera api

### PR DESCRIPTION
**Pull request by omercier**
## Description

New plugin to monitor HPE Primera storage with HTTP REST API.

REFS: CTOR-665

## Type of change

- [ ] Patch fixing an issue (non-breaking change)
- [x] New functionality (non-breaking change)
- [ ] Functionality enhancement or optimization (non-breaking change)
- [ ] Breaking change (patch or feature) that might cause side effects breaking part of the Software

## How this pull request can be tested ?

Tests provided in the PR

## Checklist

- [x] I have followed the **[coding style guidelines](https://github.com/centreon/centreon-plugins/blob/develop/doc/en/developer/plugins_global.md#5-code-style-guidelines)** provided by Centreon
- [x] I have commented my code, especially **hard-to-understand areas** of the PR.
- [ ] I have **rebased** my development branch on the base branch (develop).
- [x] I have implemented automated tests related to my commits.
- [x] I have reviewed all the help messages in all the .pm files I have modified.
  - [x] All sentences begin with a capital letter.
  - [x] All sentences are terminated by a period.
  - [x] I am able to understand all the help messages, if not, exchange with the PO or TW to rewrite them.

### CODE REVIEW

**Review by kduret — DISMISSED** [review]
You can remove `WIP` from the PR description :)

File: src/storage/hp/primera/restapi/mode/diskusage.pm [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+    };
+    $self->{disk} = {};
+
+    for my $disk (@{$disks}) {
+
+        my $disk_intro = "disk #" . $disk->{id} . " (" . $disk->{manufacturer} . "/" . $disk->{model}
+                        . ", serial: " . $disk->{serialNumber} . ") located '" . $disk->{position};
+
+        # skip if filtered by id
+        if (defined($self->{option_results}->{filter_id})
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/diskusage.pm:186]: can we put theese filter in the api.pm to factor with diskstatus.pm file ?

File: src/storage/hp/primera/restapi/mode/diskusage.pm [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+
+=head1 MODE
+
+Check disk usage.
+
+=over 8
+
+=item B<--filter-counters>
+
+Define which counters (filtered by regular expression) should be monitored.
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/diskusage.pm:263]: to standardize with the other option down here

File: src/storage/hp/primera/restapi/mode/diskusage.pm [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+
+Define which volumes should be monitored based on the disk model.
+This option will be treated as a regular expression.
+
+=item B<--filter-serial>
+
+Define which volumes should be monitored based on the disk serial number.
+This option will be treated as a regular expression.
+
+=item B<--filter-position>
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/diskusage.pm:286]: seeing the doc you wrote, should we really use a regular expression for this ?

File: src/storage/hp/primera/restapi/mode/licenses.pm [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+
+    my $response = $options{custom}->request_api(
+        endpoint => '/api/v1/system'
+    );
+    my $licenses = $response->{licenseInfo}->{licenses};
+
+    my ($total_licenses, $expired_licenses) = (0, 0);
+    for my $license_item (@{$licenses}) {
+        # skip if filter does not match
+        next if (defined($self->{option_results}->{filter_name}) and $self->{option_results}->{filter_name} ne '' and $license_item->{name} !~ /$self->{option_results}->{filter_name}/);
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/licenses.pm:150]: Line too long

File: src/storage/hp/primera/restapi/mode/listdisks.pm [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                $disk->{position},
+                $disk->{totalSizeMiB},
+                $disk->{manufacturer},
+                $disk->{model},
+                $disk->{serialNumber}
+            )
+        );
+    }
+
+    if (!defined($options{disco_show})) {
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/listdisks.pm:71]: Strange condition, what happen if the client set disco-show and the disco-format option ?

File: src/storage/hp/primera/restapi/mode/nodes.pm [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+
+    my $response = $options{custom}->request_api(
+        endpoint => '/api/v1/system'
+    );
+    my $total_nodes = 0;
+    my $online_nodes = 0;
+    my $offline_nodes = 0;
+
+    # Typical content of onlineNodes is [0, 1]
+    # %online_nodes_statuses associates 'online' to every online node is as key
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/nodes.pm:127]: to every online node is as key -> to every online node is has key ?
I don't understand

File: src/storage/hp/primera/restapi/mode/volumeusage.pm [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+=head1 MODE
+
+Check volume usage.
+
+=over 8
+
+=item B<--filter-counters>
+
+Define which counters (filtered by regular expression) should be monitored.
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/volumeusage.pm:153]: regex info should be the same format across all --filter* option (in this mode and the other)

File: src/storage/hp/primera/restapi/mode/volumeusage.pm [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+Example: --filter-counters='^usage$'
+
+=item B<--filter-id>
+
+Define which volumes should be monitored based on their IDs.
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/mode/volumeusage.pm:158]: it's not very clear if this is a white or a blacklist

File: src/storage/hp/primera/restapi/plugin.pm [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+    return $self;
+}
+
+1;
+
+__END__
+
+=head1 PLUGIN DESCRIPTION
+
+Check Control-M through Automation API.
```
Comment (Evan-Adam) [inline src/storage/hp/primera/restapi/plugin.pm:53]: copy paste ?

File: tests/robot/storage/hp/primera/restapi/licenses.robot [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    [Tags]    storage     api    hpe    hp
+    ${output}    Run    ${CMD} ${extraoptions}
+
+    ${output}    Strip String    ${output}
+
+    Should Match Regexp
+    ...    ${output}
+    ...    ${expected_result}
+    ...    Wrong output result for command:\n${CMD} ${extraoptions}\n\nObtained:\n${output}\n\nExpected:\n${expected_result}\n
+    ...    values=False
```
Comment (Evan-Adam) [inline tests/robot/storage/hp/primera/restapi/licenses.robot:36]: why don't you add colapse_space here too ?

**Review by Evan-Adam — DISMISSED** [review]

**Review by mushroomempires — APPROVED** [review]

**Review by Evan-Adam — APPROVED** [review]

**Pull request merged**